### PR TITLE
v2.0

### PR DESCRIPTION
**Pull request by kalaninja**
Hi there, 
I've managed to deal with all my "busy man" problems. So, back online and ready for some opensource.

Here is the PR for v2.0. It doesn't have a readme file yet, because I am unaware about the links to godoc, travis, etc. I assume that if we tag the new commit with as "v2.0", the link to godoc would be smth like 
https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2.0 and it seems that tha branch for travis and coveralls can be specified at their urls. Is it correct? Should I tag it myself?

### CODE REVIEW

**Comment by ahmetalpbalkan** [comment]
@kalaninja thanks for this PR! Don't worry about README etc for now, we can address the docs/website issues later.

Do you mind giving a brief overview of how much exported function/type signatures have changed? That would help me to review and give an idea of how much of a breaking change it is.

File: LICENSE [hunk 1]
```diff
@@ -1,14 +1,21 @@
-Copyright 2013 Ahmet Alp Balkan
+The MIT License (MIT)
```
Comment (ahmetalpbalkan) [inline LICENSE]: I guess this one is a leftover. Can you please revert this file?

File: example_test.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	result := []int{}
+	Range(1, 10).ToSlice(&result)
+
+	fmt.Println(result)
+
+	// Output:
+	// [1 2 3 4 5 6 7 8 9 10]
+}
+
+type Book struct {
```
Comment (ahmetalpbalkan) [inline example_test.go]: I believe this is unused

File: result.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+	for item, ok = next(); ok; item, ok = next() {
+		r += conv(item)
+	}
+
+	return
+}
+
+// ToChannel iterates over a collection and outputs each element
+// to a channel, then closes it.
+func (q Query) ToChannel(result chan<- interface{}) {
```
Comment (ahmetalpbalkan) [inline result.go]: aren't you supposed to instantiate the chan with `new`?

File: LICENSE (hunk 1 above)
Comment (kalaninja) [inline LICENSE]: Sure. Didn't notice that.

File: example_test.go (hunk 2 above)
Comment (kalaninja) [inline example_test.go]: It is an example for godoc

File: result.go (hunk 3 above)
Comment (kalaninja) [inline result.go]: The idea is that my To*Smth*() methods do not create anything, but simply populate storages created by user instead. It was necessary for the end-user in order not deal with slice of interfaces, but to work with a typed collections. And I decided to use the same principle for the ToChannel() method to keep the API easier and more understandable.

**Comment by kalaninja** [comment]
About an overview it seems to be a good idea? How do you want to have that info? I may do it as a table in excel or google docs with my methods on one side and yours on the other? Or you might have a better idea?

File: result.go (hunk 3 above)
Comment (ahmetalpbalkan) [inline result.go]: ah sorry I reviewed this 1 AM yesterday, didn't realize this is actually an input parameter. I thought it's a named return.

**Comment by ahmetalpbalkan** [comment]
@kalaninja actually, I think I got the main idea. There are quite many new methods, some methods are gone, errors are gone, `linq.T` is now `interface{}` etc.

I also found godocs to be helpful. How about you add a `README.md` (by updating the old one?) in the meanwhile and I give you commit access after merging this. The remaining tasks would be:

- we should probably add a big notice to the README.md saying v2.0 is a breaking change, for those who want, we can link to a v1.0 branch
  - for that we first need to tag a v1.0 branch :) this can be done later.
- update README with new examples
- update Wiki tab with new examples (or maybe get rid of the wiki altogether, WDYT?)
- update the `gh-pages` branch for the website with new examples

File: example_test.go (hunk 2 above)
Comment (ahmetalpbalkan) [inline example_test.go]: umm but it's not used in Example* functions either?

Comment (kalaninja) [inline example_test.go]: This function is attached to godoc automatically. You can see the result in godoc for my current repo. https://godoc.org/github.com/kalaninja/go2linq#Query.ToSlice
godoc attaches this function to Query.ToSlice() and as the output is specified in the comment it is used as a test as well.

File: result.go (hunk 3 above)
Comment (kalaninja) [inline result.go]: I understand you more that perfectly. Due to the time lag I usually answer you in the night)

**Comment by kalaninja** [comment]
Yes, I don't have error because they made the predicates look ugly, and its effect can be achieved with panic and recover.
Yes, I used interface{} instead of linq.T, but now I am unsure about that decision, probably usage of linq.T made the code look prettier. WDYT?

I'll do the readme asap. Agree with the todo list. And I do not believe that the wiki is really helpful (I didn't even know that you have one)) You have a beautiful site and we might invest more effort there.

File: example_test.go (hunk 2 above)
Comment (ahmetalpbalkan) [inline example_test.go]: Yes but I still don't see the `Book` struct anywhere?

File: compare.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+type comparer func(interface{}, interface{}) int
+
+// Comparable is an interface that has to be implemented by a
+// custom collection elememts in order to work with linq.
+//
+// Example:
+//		func (f foo) CompareTo(c Comparable) int {
+//			a, b := f.f1, c.(foo).f1
+//
+//			if a < b {
```
Comment (ahmetalpbalkan) [inline compare.go]: this code would be shorter as

```
switch{
case a>b: return -1
case b<a: return 1
default: return 0
}
```

File: example_test.go (hunk 2 above)
Comment (kalaninja) [inline example_test.go]: Oh, I see! I probably wanted to make some example... Now this struct seems abandoned, I'll remove it.

**Comment by ahmetalpbalkan** [comment]
@kalaninja This generally looks good to me. Can you just add a README by copying from the current one and updating the example etc. Then it would be good to merge. Then we can address other stuff in https://github.com/ahmetalpbalkan/go-linq/pull/30#issuecomment-243571343.

File: compare.go (hunk 4 above)
Comment (kalaninja) [inline compare.go]: they seem pretty the same legth since I use gofmt
			switch {
			case a > b:
				return -1
			case b < a:
				return 1
			default:
				return 0
			}
vs
			if a < b {
				return -1
			} else if a > b {
				return 1
			}

			return 0

There are other variants as well. Consider:
			if a < b {
				return -1
			}
			if a > b {
				return 1
			}
			return 0
WDYT?

Comment (ahmetalpbalkan) [inline compare.go]: I guess it doesn't matter too much then. I still prefer

```
	switch {
	case a > b:
		return -1
	case b < a:
		return 1
	default:
		return 0
	}
```

mostly because it's easier to follow

Comment (kalaninja) [inline compare.go]: Yes, I agree. It looks really straightforward. I'll change that.

**Comment by kalaninja** [comment]
I've attached the readme, changed "if" to "switch" and removed the book struct from the example.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	}, func(a interface{}) interface{} {
+		return a
+	}).OrderByDescending(func(g interface{}) interface{} {
+		return len(g.(Group).Group)
+	}).Select(func(g interface{}) interface{} {
+		return g.(Group).Key
+	}).First()
+```
+
+**More examples** can be found in [documentation](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2)
```
Comment (ahmetalpbalkan) [inline README.md]: Maybe don't link to these like v2 yet because we'll merge it to `master`.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+From(cars).Where(func(c interface{}) bool {
+	return c.(Car).year >= 2015
+}).Select(func(c interface{}) interface{} {
+	return c.(Car).owner
+}).ToSlice(&owners)
+```
+
+**Example:** Find an author that has written the most books
+```go
+import . "gopkg.in/ahmetalpbalkan/go-linq.v2"
```
Comment (ahmetalpbalkan) [inline README.md]: We shouldn't suggest using gopkg.in.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
+# go-linq [![GoDoc](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2?status.svg)](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2) [![Build Status](https://travis-ci.org/ahmetalpbalkan/go-linq.svg?branch=v2.0)](https://travis-ci.org/ahmetalpbalkan/go-linq) [![Coverage Status](https://coveralls.io/repos/github/ahmetalpbalkan/go-linq/badge.svg?branch=v2.0)](https://coveralls.io/github/ahmetalpbalkan/go-linq?branch=v2.0)
+A powerful language integrated query (LINQ) library for Go.
+* Safe for concurrent use
+* Complete lazy evaluation
+* Supports arrays, slices, maps, strings, channels and
+custom collections (collection needs to implement Iterable interface
+and element - Comparable interface)
+* Parallel LINQ (PLINQ) *(comming soon)*
```
Comment (ahmetalpbalkan) [inline README.md]: hmm maybe don't mention this just yet?

File: README.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
+# go-linq [![GoDoc](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2?status.svg)](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2) [![Build Status](https://travis-ci.org/ahmetalpbalkan/go-linq.svg?branch=v2.0)](https://travis-ci.org/ahmetalpbalkan/go-linq) [![Coverage Status](https://coveralls.io/repos/github/ahmetalpbalkan/go-linq/badge.svg?branch=v2.0)](https://coveralls.io/github/ahmetalpbalkan/go-linq?branch=v2.0)
+A powerful language integrated query (LINQ) library for Go.
```
Comment (ahmetalpbalkan) [inline README.md]: I see you removed the notice about **written in vanilla Go**, doesn't this still hold true?

File: README.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
+# go-linq [![GoDoc](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2?status.svg)](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2) [![Build Status](https://travis-ci.org/ahmetalpbalkan/go-linq.svg?branch=v2.0)](https://travis-ci.org/ahmetalpbalkan/go-linq) [![Coverage Status](https://coveralls.io/repos/github/ahmetalpbalkan/go-linq/badge.svg?branch=v2.0)](https://coveralls.io/github/ahmetalpbalkan/go-linq?branch=v2.0)
```
Comment (ahmetalpbalkan) [inline README.md]: Hmm let's just remove all v2.0s here. It's not going to be a branch, I'm planning to make this live in master (and tag it as `v2.0` obviously) and archive the old version in a branch named `archive/0.9` if we need to take any fixes in the old code.

This code will be `master`, so we don't need to specify a branch name.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+* Safe for concurrent use
+* Complete lazy evaluation
+* Supports arrays, slices, maps, strings, channels and
+custom collections (collection needs to implement Iterable interface
+and element - Comparable interface)
+* Parallel LINQ (PLINQ) *(comming soon)*
+
+## Installation
+
+    $ go get gopkg.in/ahmetalpbalkan/go-linq.v2
```
Comment (ahmetalpbalkan) [inline README.md]: Let's just use github here as well. I'm in the camp of people who are against pulling from gopkg.in as well. It's more likely that github will last longer :)

File: README.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+```go
+import . "gopkg.in/ahmetalpbalkan/go-linq.v2"
+	
+type Book struct {
+	id      int
+	title   string
+	authors []string
+}
+
+author := From(books).SelectMany(func(b interface{}) Query {
```
Comment (ahmetalpbalkan) [inline README.md]: I find it a bit hard to follow this example 😵 . Not sure if it's just me but all the `a`s and `b`s kind of seemed confusing.

Comment (ahmetalpbalkan) [inline README.md]: I've been thinking how to make this more understandable. Perhaps do not follow gofmt indentation rules and squeeze some funcs into single lines and do a custom indentation, and add some `// comments`?

File: README.md (hunk 5 above)
Comment (kalaninja) [inline README.md]: And what will happen with the current version?

File: README.md (hunk 7 above)
Comment (kalaninja) [inline README.md]: Oh, sure. I also thought to delete this. Especially, because similar functionality can be achieved From(channel) and ToChannel() methods

File: README.md (hunk 6 above)
Comment (kalaninja) [inline README.md]: ok. I didn't know that

File: README.md (hunk 8 above)
Comment (kalaninja) [inline README.md]: Yes, that is still true, and sounds really cool. I'll bring it back

File: README.md (hunk 9 above)
Comment (kalaninja) [inline README.md]: Ok, I got that

File: README.md (hunk 10 above)
Comment (kalaninja) [inline README.md]: Sounds reasonable. I'll fix everything connected with branching

File: README.md (hunk 5 above)
Comment (ahmetalpbalkan) [inline README.md]: I archived the current version in `archive/0.9` branch, anybody who wants to use it I guess can grab it from there or perhaps grab it via `go get gopkg.in/...` (we have a `v0.9-rc5` tag, we can tag it as `v0.9` anytime if that'll make it work with gopkg.in).

File: README.md [hunk 12]
```diff
@@ -1,100 +1,63 @@
...
-author's employment association with Microsoft does not imply any sort of
-warranty or official representation the company. This is purely a personal side
-project developed on spare times.
+owners := []string{}
 
-## Authors
-
-[Ahmet Alp Balkan](http://ahmetalpbalkan.com) – [@ahmetalpbalkan](https://twitter.com/ahmetalpbalkan)
-
-## Release Notes
```
Comment (ahmetalpbalkan) [inline README.md]: can you please  keep the Release Notes in the new README? and also add v2.0 briefly stating it is a breaking change and a total rewrite with greater performance, efficiency and new functionality?

Comment (kalaninja) [inline README.md]: Do you want to keep all the previous notes?

Comment (kalaninja) [inline README.md]: I'll put the release notes in the beginning, since for the the guys that already use go-linq  it is probably more important than the examples

Comment (ahmetalpbalkan) [inline README.md]: Hmm you can leave it here. I was planning to add a big notice to the top about the breaking change after merging this.

File: README.md (hunk 11 above)
Comment (kalaninja) [inline README.md]: does it look better?

**Comment by kalaninja** [comment]
I've fixed links in readme, removed the line about plinq, recovered the one about vanilla go, added release notes.
I've tried several ways to make the example more readable and ended with renaming a to author, b to book and g to group. I've added comments as well.

**Comment by ahmetalpbalkan** [comment]
@kalaninja thanks! I'll open issues for the rest of the updates we probably need to make.

**Comment by ahmetalpbalkan** [comment]
💥 🎊 🍷
